### PR TITLE
new d3d12 wrapping related work and fixes

### PR DESCRIPTION
**Pull request by galop1n**
First pull request attempt, i am not really used to git and i don't know the way to keep the visual project updates local ( as i work in 2015 ) and not send them in the pull request.

Update to latest d3d12 interface ( for ID3D12Device1 and ID3D12Debug1 )
Fixes of missing and improper unwrap.
Handle wrapping of the remaining d3d12 interface creation calls.
Fix of d3d12 reusing root signature pointer at creation.
Fix read back resource creation at capture time

All the files were processed by clang-format before commits.

### CODE REVIEW

File: renderdoc/driver/d3d12/d3d12_command_list_wrap.cpp [hunk 1]
```diff
@@ -812,11 +812,33 @@ void WrappedID3D12GraphicsCommandList::ResourceBarrier(UINT NumBarriers,
...
-    barriers[i].Transition.pResource = Unwrap(barriers[i].Transition.pResource);
+    if(barriers[i].Type == D3D12_RESOURCE_BARRIER_TYPE_TRANSITION)
+    {
+      barriers[i].Transition.pResource = Unwrap(barriers[i].Transition.pResource);
 
-    // hack while not all resources are wrapped
-    if(barriers[i].Transition.pResource == NULL)
-      barriers[i].Transition.pResource = pBarriers[i].Transition.pResource;
+      // hack while not all resources are wrapped
+      if(barriers[i].Transition.pResource == NULL)
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_command_list_wrap.cpp]: This hack can be deleted from all paths now, this was only while I was doing early testing and hadn't wrapped some simple things yet. There are still some unwrapped resources but they're not needed for a 'hello world' type application, so it's OK to fail here - they just need to be implemented.

File: renderdoc/driver/d3d12/d3d12_manager.cpp [hunk 2]
```diff
@@ -373,12 +373,8 @@ bool D3D12ResourceManager::Prepare_InitialState(ID3D12DeviceChild *res)
...
     if(desc.Dimension == D3D12_RESOURCE_DIMENSION_BUFFER)
     {
-      D3D12_HEAP_PROPERTIES heapProps;
-      heapProps.Type = D3D12_HEAP_TYPE_READBACK;
-      heapProps.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
-      heapProps.MemoryPoolPreference = D3D12_MEMORY_POOL_UNKNOWN;
-      heapProps.CreationNodeMask = 1;
-      heapProps.VisibleNodeMask = 1;
+      D3D12_HEAP_PROPERTIES heapProps =
+          m_Device->GetReal()->GetCustomHeapProperties(1, D3D12_HEAP_TYPE_READBACK);
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_manager.cpp]: What's the purpose of this change? My understanding is that `GetCustomHeapProperties` returns an entirely equivalent set of heap properties to using the `READBACK` macro.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 3]
```diff
@@ -728,26 +901,174 @@ HRESULT WrappedID3D12Device::CreateCommittedResource(const D3D12_HEAP_PROPERTIES
...
+      Serialise_CreateHeap(pDesc, riid, (void **)&wrapped);
+
+      D3D12ResourceRecord *record = GetResourceManager()->AddResourceRecord(wrapped->GetResourceID());
+      record->type = Resource_Heap;
+      record->Length = 0;
+      wrapped->SetResourceRecord(record);
+
+      record->AddChunk(scope.Get());
+
+      {
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: I'm not 100% sure yet but I don't think heaps should be dirty-tracked, I think it will be better just to track individual resources.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 4]
```diff
@@ -660,6 +660,179 @@ bool WrappedID3D12Device::Serialise_CreateCommittedResource(
   return true;
 }
 
+bool WrappedID3D12Device::Serialise_CreatePlacedResource(ID3D12Heap *pHeap, UINT64 HeapOffset,
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: The `Serialise_` functions should always be placed immediately above the function they're serialising.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 5]
```diff
@@ -728,26 +901,174 @@ HRESULT WrappedID3D12Device::CreateCommittedResource(const D3D12_HEAP_PROPERTIES
...
+
+  if(SUCCEEDED(ret))
+  {
+    SCOPED_LOCK(m_D3DLock);
+
+    WrappedID3D12Resource *wrapped = new WrappedID3D12Resource(real, this);
+
+    if(m_State >= WRITING)
+    {
+      SCOPED_SERIALISE_CONTEXT(CREATE_COMMITTED_RESOURCE);
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: Should be `CREATE_RESERVED_RESOURCE` here.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 6]
```diff
@@ -822,14 +1143,101 @@ HRESULT WrappedID3D12Device::CreateFence(UINT64 InitialValue, D3D12_FENCE_FLAGS
...
+      Serialise_CreateQueryHeap(pDesc, riid, (void **)&wrapped);
+
+      D3D12ResourceRecord *record = GetResourceManager()->AddResourceRecord(wrapped->GetResourceID());
+      record->type = Resource_QueryHeap;
+      record->Length = 0;
+      wrapped->SetResourceRecord(record);
+
+      record->AddChunk(scope.Get());
+
+      {
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: Same again, I think query heaps definitely shouldn't be dirty-tracked. I'm not even sure it would be feasible to fetch and restore their state..

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 7]
```diff
@@ -728,26 +901,174 @@ HRESULT WrappedID3D12Device::CreateCommittedResource(const D3D12_HEAP_PROPERTIES
 
 HRESULT WrappedID3D12Device::CreateHeap(const D3D12_HEAP_DESC *pDesc, REFIID riid, void **ppvHeap)
 {
-  return m_pDevice->CreateHeap(pDesc, riid, ppvHeap);
+  if(ppvHeap == NULL)
+    return m_pDevice->CreateHeap(pDesc, riid, nullptr);
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: Just use `NULL` here, not `nullptr`, also in the same place for in the other functions added.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 8]
```diff
@@ -822,14 +1143,101 @@ HRESULT WrappedID3D12Device::CreateFence(UINT64 InitialValue, D3D12_FENCE_FLAGS
...
+      Serialise_CreateCommandSignature(pDesc, pRootSignature, riid, (void **)&wrapped);
+
+      D3D12ResourceRecord *record = GetResourceManager()->AddResourceRecord(wrapped->GetResourceID());
+      record->type = Resource_CommandSignature;
+      record->Length = 0;
+      wrapped->SetResourceRecord(record);
+
+      record->AddChunk(scope.Get());
+
+      {
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: Ditto here, command signatures have no state so shouldn't be marked dirty.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 9]
```diff
@@ -1266,7 +1266,7 @@ void WrappedID3D12Device::CopyDescriptors(
                              NumSrcDescriptorRanges, srcStarts, pSrcDescriptorRangeSizes,
                              DescriptorHeapsType);
 
-  RDCUNIMPLEMENTED("CopyDescriptors does not copy our internal descriptor data");
+	D3D12NOTIMP("CopyDescriptors does not copy our internal descriptor data");
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: Yes this was just leftover by accident when I implemented the function, it can be deleted! :smile:

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 10]
```diff
@@ -582,7 +582,9 @@ void WrappedID3D12Device::CreateShaderResourceView(ID3D12Resource *pResource,
                                                    D3D12_CPU_DESCRIPTOR_HANDLE DestDescriptor)
 {
   GetWrapped(DestDescriptor)->Init(pResource, pDesc);
-  return m_pDevice->CreateShaderResourceView(Unwrap(pResource), pDesc, Unwrap(DestDescriptor));
+  if(!pResource)
+    return;
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: I guess this is coming from `D3D12Descriptor::Create` where we create a SRV with a NULL resource for undefined descriptors?

I think we'll need those to be well-defined empty descriptors rather than invalid uninitialised ones. I think the proper fix here would be to determine the heap type and create the appropriate NULL resource descriptor (since it doesn't matter if it's SRV or what, just that it's valid).

I believe doing this early return will prevent any application-created descriptors with a NULL resource from initializing, which could lead to crashes when they're used in rendering.

File: renderdoc/driver/d3d12/d3d12_device_wrap.cpp [hunk 11]
```diff
@@ -747,7 +755,7 @@ bool WrappedID3D12Device::Serialise_CreateReservedResource(
 bool WrappedID3D12Device::Serialise_CreateHeap(const D3D12_HEAP_DESC *pHeapDesc,
                                                REFIID riidResource, void **ppvResource)
 {
-  SERIALISE_ELEMENT(D3D12_HEAP_DESC, desc, *pHeapDesc);
+  SERIALISE_ELEMENT_PTR(D3D12_HEAP_DESC, desc, pHeapDesc);
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_device_wrap.cpp]: FWIW I think this change is redundant but fine to make. The `SERIALISE_ELEMENT_PTR` is kind of leftovers, and would only really be useful if you had some pointer that could be NULL and you wanted a default-initialised value instead. Here, `pHeapDesc` is required.

**Comment by baldurk** [comment]
A couple of bookkeeping things - yes you'll need to remove the project and header changes from the pull request - there are a few ways to do this, one would be to work on a given scratch branch, and then when you're ready to make your PR create a new branch, and rebase it to remove those commits. This would also let you remove the clang-format commits which I'd like you to please squash into other commits as appropriate - it leads to an ugly history if they're left separate.

On that note - please don't add new commits to a PR after you've created it, it makes things harder to process. The original PR was an OK size but this is now a bit of a clump of loosely related changes, ideally you'd make a PR with a particular purpose since it's then easier to track in history as well as review.

In more general terms, the D3D12 back-end is still in its very early stages which is why I've not really announced its existance much. I'm grateful for help and I'm not going to reject any PRs over it, but I'd kind of suggest you stay away from making changes here as there's going to be an awful lot of churn that will make things difficult.

Because of the limitations of D3D12 needing Windows 10, I can't develop it on my primary machine, I have to use a secondary PC that runs windows 10. That sometimes means that I don't push my work up to github as regularly, so you might end up doing work that conflicts with stuff I have done - and I am likely to accept mine rather than yours in that case. Again I appreciate the contribution I just want to make sure there's not going to be conflicts and such wasting work.

**Comment by galop1n** [comment]
Hi,

First, thanks for the time spent reviewing that PR. I will process everything in the comments, break up in smaller PR, clean up, then come back to bother you again. I definitely have perforce habits that does not mix well with a git flow i need to work on.

I will certainly continue to work on the d3d12 backend, because VSGD is crap, IHV tools inconsistent, and PIX Xbox One only. And Renderdoc was an important tool for us with DX11 and it will likely be for DX12 too, building a knowledge of his code is a safe guard to help internally fix future bug if needed.

I am fine with you not having the time to review all the crap i will commit, or picking your own code if it conflicts with something you have locally and not yet on github. I don't know how it will work with future pull request as i move forward tho.

Right now, i can run and capture my dx12 sandbox ( mostly every piece of api possible ), so it is definitely an improvement ( i need the reference all resource check box and loading the capture crash on a failing CreateShaderResourceView ).

Thanks.

**Comment by baldurk** [comment]
No problem. I think there are some good git process guides out there. People have different thoughts and opinions on git, but I like the workflow and how flexible it can be. E.g. it's very easy to keep a branch somewhere, and make a new branch with only a few of the commits (look up `git cherry-pick` to bring across a single commit). I also make extensive use of `git rebase -i` in my work. E.g. when I was doing some of the D3D12 work I would make many commits, find a bug or mistake in one of them later, and then I would just make a new commit "fix to <x> earlier thing", and then `git rebase -i` allows me to both re-order them so the fix is immediately afterwards, and then also squash them together. This could be useful both for some of the bugfix commits you have as well as the clang-format commits. If you want any git help, get in touch directly and I can give you pointers.

I'm glad you have a high opinion of RenderDoc, hopefully the D3D12 support can live up to that :smile:. As I say you are of course free to continue working on it and I'll accept contributions, I just wanted to make sure your expectations are set, since with the code still largely unfinished and in flux there's a higher chance of me just going "this isn't correct I'm going to do it a different way", or running into conflicts that either mean your work won't be applied or at least will have to be resolved against my changes.

I'll definitely review anything you submit though and pull in what I can (right now I don't have any pending work so this can all go in). I'll also try and do better at committing the D3D12 work more regularly so you can see the latest. Another git thing - when you make a pull request, you'll need to make sure it is rebased against recent code, rather than merging.

Good news about the sandbox though! I've been using [AMD's HelloD3D12](https://github.com/GPUOpen-LibrariesAndSDKs/HelloD3D12) for testing as it's about as simple as it gets, I didn't realise the API coverage was remotely useful for something more complex, but missing capture data might not show up as a crash, so it won't be clear until you can replay :smile:.

I'm closing this PR since you'll split it up, but again - the code contributions are welcome.

**Comment by oscarbg** [comment]
Hi,
I compiled Renderdoc D3D12 from galop1n branch which should be equal to using this patches and using Microsoft  DirectX12 samples from(https://github.com/Microsoft/DirectX-Graphics-Samples).
I tested simple D3D12HelloTriangle from D3D12HelloWorld solution..
I compiled a x86 binary of Renderdoc and of the sample and I could record a frame but upon replaying crashes.. 
Replay Log:
RENDERDOC:  [09:05:17]             core.cpp( 270) - Log     - RenderDoc v0.31 x86 (NO_GIT_COMMIT_HASH_DEFINED) loaded in replay application
RENDERDOC:  [09:05:43]    win32_process.cpp( 410) - Log     - Injecting renderdoc into process 14912
RENDERDOC:  [09:05:43]    win32_process.cpp( 279) - Debug   - Injecting call to RENDERDOC_SetLogFile
RENDERDOC:  [09:05:43]    win32_process.cpp( 279) - Debug   - Injecting call to RENDERDOC_SetCaptureOptions
RENDERDOC:  [09:05:43]    win32_process.cpp( 279) - Debug   - Injecting call to RENDERDOC_GetRemoteAccessIdent
RENDERDOC:  [09:05:43]    win32_network.cpp( 339) - Debug   - connect before timeout
RENDERDOC:  [09:05:43]    remote_access.cpp( 398) - Log     - Got remote handshake: D3D12HelloTriangle () [14912]
RENDERDOC:  [09:05:44]    remote_access.cpp( 600) - Log     - Used API: D3D12
RENDERDOC:  [09:05:46]    remote_access.cpp( 587) - Log     - Got a new capture: 0 (time 1472367946) 0 byte thumbnail
RENDERDOC:  [09:05:48]    win32_network.cpp( 178) - Warning - recv: 10054
RENDERDOC:  [09:05:48]       serialiser.cpp( 629) - Debug   - Opened capture file for read
RENDERDOC:  [09:05:48]  replay_renderer.cpp(1551) - Log     - Creating replay device for C:\Users\oscarb\AppData\Local\Temp\D3D12HelloTriangle_2016.08.28_09.05.43_frame98.rdc
RENDERDOC:  [09:05:48]       serialiser.cpp( 629) - Debug   - Opened capture file for read
RENDERDOC:  [09:05:48]     d3d12_replay.cpp( 522) - Debug   - Creating a D3D12 replay device
RENDERDOC:  [09:05:48]       serialiser.cpp( 629) - Debug   - Opened capture file for read
RENDERDOC:  [09:05:48]      d3d12_hooks.cpp( 173) - Debug   - Call to Create_Internal Feature Level b000
RENDERDOC:  [09:05:48]      d3d12_hooks.cpp( 179) - Debug   - In replay app
RENDERDOC:  [09:05:48]      d3d12_hooks.cpp( 214) - Debug   - Calling real createdevice...
RENDERDOC:  [09:05:49]      d3d12_hooks.cpp( 235) - Debug   - Called real createdevice... 0x00000000
RENDERDOC:  [09:05:49]      d3d12_hooks.cpp( 239) - Debug   - succeeded and hooking.
RENDERDOC:  [09:05:49]      d3d12_hooks.cpp( 250) - Debug   - created wrapped device.
RENDERDOC:  [09:05:49]       serialiser.cpp( 629) - Debug   - Opened capture file for read
RENDERDOC:  [09:05:49]     d3d12_replay.cpp( 581) - Log     - Created device.
RENDERDOC:  [09:05:49]  replay_renderer.cpp(1568) - Log     - Created replay driver.
RENDERDOC:  [09:05:49]     d3d12_common.cpp( 800) - Error   - Unrecognised RTV Dimension 0
RENDERDOC:  [09:05:49]     d3d12_common.cpp( 800) - Error   - Unrecognised RTV Dimension 0
RENDERDOC:  [09:05:49]     d3d12_common.cpp( 752) - Error   - Unrecognised SRV Dimension 0
RENDERDOC:  [09:05:49]    d3d12_manager.cpp( 647) - Error   - Couldn't create upload buffer: 0x8007000e
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 12 at offset 157067
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 225 at offset 157241
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 4592 at offset 157259
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: f3
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 12676 at offset 157265
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 7710 at offset 158203
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 17
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 6977 at offset 158209
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 7926 at offset 158218
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 7213 at offset 158311
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 8866 at offset 158358
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 255 at offset 158914
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 2316 at offset 158920
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 206 at offset 158926
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 2
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 10477 at offset 158932
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 256 at offset 158940
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 89 at offset 158946
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 111 at offset 158952
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 11776 at offset 158958
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 40
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 8445 at offset 158962
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 16221 at offset 158977
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 1293 at offset 159518
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 512 at offset 161539
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 616 at offset 161665
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 7
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 545 at offset 161669
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 27
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 8039 at offset 162124
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 6144 at offset 162131
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 2e
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 10599 at offset 162193
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 10087 at offset 162202
RENDERDOC:  [09:05:49] d3d12_device_wrap.cpp( 422) - Error   - Failed on resource serialise-creation, HRESULT: 0x80070057
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 38 at offset 162252
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 12166 at offset 162258
RENDERDOC:  [09:05:49]       serialiser.cpp(1587) - Error   - Unexpected control byte: 40
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 10087 at offset 162262
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 50 at offset 162269
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 4612 at offset 162275
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 36 at offset 162279
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 10087 at offset 162285
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 39 at offset 162289
RENDERDOC:  [09:05:49]     d3d12_device.cpp(1459) - Error   - Unexpected non-device chunk 512 at offset 162295
RENDERDOC:  [09:05:49]       serialiser.cpp(1124) - Fatal   - Allocation for 1734814784 bytes failed


Capture Log:
RENDERDOC:  [09:05:43]             core.cpp( 273) - Log     - RenderDoc v0.31 x86 (NO_GIT_COMMIT_HASH_DEFINED) capturing application
RENDERDOC:  [09:05:43]   win32_libentry.cpp(  66) - Log     - Loading into C:\Users\oscarb\Downloads\DirectX-Graphics-Samples-1.0.160510.1\DirectX-Graphics-Samples-1.0.160510.1\Samples\Desktop\D3D12HelloWorld\src\HelloTriangle\bin\Win32\Release\D3D12HelloTriangle.exe
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook kernel32.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into kernel32.dll, PID 14912
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook opengl32.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into opengl32.dll, PID 14912
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook d3d11.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into d3d11.dll, PID 14912
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook d3d9.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into d3d9.dll, PID 14912
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook dxgi.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into dxgi.dll, PID 14912
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook d3d12.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into d3d12.dll, PID 14912
RENDERDOC:  [09:05:43]            hooks.cpp(  45) - Debug   - Attempting to hook vulkan-1.dll
RENDERDOC:  [09:05:43]            hooks.cpp(  49) - Log     - Loaded and hooked into vulkan-1.dll, PID 14912
RENDERDOC:  [09:05:43]      d3d12_hooks.cpp( 173) - Debug   - Call to Create_Internal Feature Level b000
RENDERDOC:  [09:05:43]      d3d12_hooks.cpp( 214) - Debug   - Calling real createdevice...
RENDERDOC:  [09:05:43]     dxgi_wrapped.cpp( 518) - Warning - Querying IDXGIObject for interface: GUID {7abb6563-02bc-47c4-8ef9-acc4795edbcf}
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 235) - Debug   - Called real createdevice... 0x00000001
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 257) - Log     - Created wrapped D3D12 device.
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 173) - Debug   - Call to Create_Internal Feature Level b000
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 214) - Debug   - Calling real createdevice...
RENDERDOC:  [09:05:44]     dxgi_wrapped.cpp( 518) - Warning - Querying IDXGIObject for interface: GUID {7abb6563-02bc-47c4-8ef9-acc4795edbcf}
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 235) - Debug   - Called real createdevice... 0x00000000
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 239) - Debug   - succeeded and hooking.
RENDERDOC:  [09:05:44]      d3d12_hooks.cpp( 250) - Debug   - created wrapped device.
RENDERDOC:  [09:05:44]    d3d12_resources.h( 569) - Debug   - D3D12 not implemented - Resource mapping
RENDERDOC:  [09:05:44]    d3d12_resources.h( 575) - Debug   - D3D12 not implemented - Resource mapping
RENDERDOC:  [09:05:44]     d3d12_device.cpp(1168) - Debug   - D3D12 not implemented - ReleaseResource
RENDERDOC:  [09:05:46]   resource_manager.h( 896) - Debug   - Preparing up to 32 potentially dirty resources
RENDERDOC:  [09:05:46]   resource_manager.h( 921) - Debug   - Prepared 31 dirty resources
RENDERDOC:  [09:05:46]   resource_manager.h( 937) - Debug   - Force-prepared 0 dirty resources
RENDERDOC:  [09:05:46]     d3d12_device.cpp( 974) - Debug   - Attempting capture
RENDERDOC:  [09:05:46]     d3d12_device.cpp( 990) - Log     - Starting capture, frame 97
RENDERDOC:  [09:05:46] d3d12_command_queue_wrap.cpp(  83) - Debug   - D3D12 not implemented - Serialise_DebugMessages
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1023) - Log     - Finished capture, Frame 98
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1087) - Debug   - Inserting Resource Serialisers
RENDERDOC:  [09:05:46]   resource_manager.h( 859) - Debug   - 7 frame resource records
RENDERDOC:  [09:05:46]   resource_manager.h( 881) - Debug   - 46 frame resource chunks
RENDERDOC:  [09:05:46]   resource_manager.h( 888) - Debug   - inserted to serialiser
RENDERDOC:  [09:05:46]   resource_manager.h( 949) - Debug   - Checking 32 possibly dirty resources
RENDERDOC:  [09:05:46]     d3d12_common.cpp( 800) - Error   - Unrecognised RTV Dimension 0
RENDERDOC:  [09:05:46]     d3d12_common.cpp( 800) - Error   - Unrecognised RTV Dimension 0
RENDERDOC:  [09:05:46]     d3d12_common.cpp( 752) - Error   - Unrecognised SRV Dimension 0
RENDERDOC:  [09:05:46]     d3d12_common.cpp( 800) - Error   - Unrecognised RTV Dimension 0
RENDERDOC:  [09:05:46]     d3d12_common.cpp( 800) - Error   - Unrecognised RTV Dimension 0
RENDERDOC:  [09:05:46]   resource_manager.h(1027) - Debug   - Serialised 31 dirty resources, skipped 0 unreferenced
RENDERDOC:  [09:05:46]   resource_manager.h(1058) - Debug   - Force-serialised 0 dirty resources
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1093) - Debug   - Creating Capture Scope
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1114) - Debug   - Flushing 1 command list records to file serialiser
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1121) - Debug   - Adding 11 chunks to file serialiser from command list 302
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1131) - Debug   - Flushing 14 chunks to file serialiser from context record
RENDERDOC:  [09:05:46]     d3d12_device.cpp(1136) - Debug   - Done
RENDERDOC:  [09:05:46]       serialiser.cpp(1277) - Debug   - writing capture files
RENDERDOC:  [09:05:46]       serialiser.cpp(1305) - Debug   - Opened capture file for write
RENDERDOC:  [09:05:46]       serialiser.cpp(1415) - Log     - Compressed frame capture data from 334002 to 49287
RENDERDOC:  [09:05:46]       serialiser.cpp(1273) - Log     - Timer File writing - 17.594 ms
RENDERDOC:  [09:05:46]             core.cpp( 858) - Log     - Written to disk: C:\Users\oscarb\AppData\Local\Temp\D3D12HelloTriangle_2016.08.28_09.05.43_frame98.rdc
RENDERDOC:  [09:05:46]       serialiser.cpp( 629) - Debug   - Opened capture file for read
RENDERDOC:  [09:05:46]       serialiser.cpp( 629) - Debug   - Opened capture file for read
RENDERDOC:  [09:05:48]      d3d12_debug.cpp(1199) - Debug   - D3D12 not implemented - Not freeing RTV's - will run out

**Comment by galop1n** [comment]
@oscarbg It is also the issue i have with a much more complex and real life sample, if i read the log properly ( RTV failure ). I believe it is something wrong in the capture, not in the replay. I started working locally on @baldurk remarks, but it had been a slow week for side projects.

Thanks for showing interests in my branch tho
